### PR TITLE
Convert JSON graph data into SVG DOM elements, and populate canvas with them.

### PR DESCRIPTION
**Pull request by R0ss**

### CODE REVIEW

File: public/js/draw/draw.js [hunk 1]
```diff
@@ -389,3 +389,99 @@ function regionClicked(e) {
...
+        node.setAttribute('x', jsonShape.pos[0]);
+        node.setAttribute('y', jsonShape.pos[1]);
+        node.setAttribute('rx', 4);
+        node.setAttribute('ry', 4);
+        node.setAttribute('id', 'n' + nodeId);
+        node.setAttribute('width', jsonShape.width);
+        node.setAttribute('height', jsonShape.height);
+        node.setAttribute('class', 'node');
+        node.parents = [];
+        node.kids = [];
```
Comment (Ian-Stewart-Binks) [inline public/js/draw/draw.js]: I realize that `kids` is a relic from the creation of the draw tool, but is there a better name for it? Perhaps a name that doesn't necessitate a comment?

File: public/js/draw/draw.js [hunk 2]
```diff
@@ -389,3 +389,99 @@ function regionClicked(e) {
...
 }
+
+
+/**
+ * Convert JSON data into SVG elements then populate the canvas with them.
+ */
+function jsonToSvg(jsonStr) {
+    'use strict';
+
+    var jsonData = JSON.parse(jsonStr);
```
Comment (Ian-Stewart-Binks) [inline public/js/draw/draw.js]: @david-yz-liu I noticed that arrows have been introduced in #805. Do we have plans to move to `let` over `var`?

File: public/js/draw/draw.js [hunk 3]
```diff
@@ -389,3 +389,99 @@ function regionClicked(e) {
...
+        var textNode = document.createTextNode(jsonShape.text[0].text);
+        code.appendChild(textNode);
+        g.appendChild(code);
+
+        document.getElementById('t' + nodeId).addEventListener('mousedown', nodeClicked, false);
+
+        nodeId += 1;
+    });
+
+    // Parse Text data
```
Comment (Ian-Stewart-Binks) [inline public/js/draw/draw.js]: I think that these comments signal that these blocks can be broken off into a separate function. @R0ss What do you think?

File: public/js/draw/draw.js [hunk 4]
```diff
@@ -389,3 +389,99 @@ function regionClicked(e) {
...
+        node.setAttribute('y', jsonShape.pos[1]);
+        node.setAttribute('rx', 4);
+        node.setAttribute('ry', 4);
+        node.setAttribute('id', 'n' + nodeId);
+        node.setAttribute('width', jsonShape.width);
+        node.setAttribute('height', jsonShape.height);
+        node.setAttribute('class', 'node');
+        node.parents = [];
+        node.kids = [];
+        // note: children doesn't work because javascript objects already have a children attribute
```
Comment (Ian-Stewart-Binks) [inline public/js/draw/draw.js]: This comment does not need to be in the codebase twice.

File: public/js/draw/draw.js [hunk 5]
```diff
@@ -389,3 +389,99 @@ function regionClicked(e) {
...
+
+        g.setAttribute('class', 'node');
+        g.setAttribute('id', 'g' + nodeId);
+        g.setAttribute('data-active', 'active');
+        g.setAttribute('fill', jsonShape.fill);
+
+        node.setAttribute('x', jsonShape.pos[0]);
+        node.setAttribute('y', jsonShape.pos[1]);
+        node.setAttribute('rx', 4);
+        node.setAttribute('ry', 4);
```
Comment (Ian-Stewart-Binks) [inline public/js/draw/draw.js]: How much of this code is duplicated in this file?

File: hs/Response/Loading.hs [hunk 6]
```diff
@@ -32,5 +32,4 @@ largeLoadingIcon = H.div ! A.id "loading-icon" $ do
 smallLoadingIcon :: H.Html
 smallLoadingIcon = H.div ! A.id "loading-icon" $ do
               H.img ! A.id "c-logo-small" ! A.src "static/res/img/C-logo-small.png"
-              H.img ! A.id "compass-small" ! A.class_ "spinner" ! A.src "static/res/img/compass-small.png"
-
```
Comment (Ian-Stewart-Binks) [inline hs/Response/Loading.hs]: This PR doesn't need to be built on top of #807. They are independent.

File: public/js/draw/draw.js [hunk 7]
```diff
@@ -389,3 +389,99 @@ function regionClicked(e) {
         prevY = position.y;
     }
 }
+
+
+/**
+ * Convert JSON data into SVG elements then populate the canvas with them.
+ */
+function jsonToSvg(jsonStr) {
```
Comment (Ian-Stewart-Binks) [inline public/js/draw/draw.js]: This function does two things, but the function name only represents one of them. This function name makes me think this just converts a JSON string to an SVG string, but this is not the case.

File: public/js/draw/draw.js (hunk 1 above)
Comment (david-yz-liu) [inline public/js/draw/draw.js]: The name is fine.

File: public/js/draw/draw.js (hunk 2 above)
Comment (david-yz-liu) [inline public/js/draw/draw.js]: Eventually, but not now.

**Comment by david-yz-liu** [comment]
It seems ellipses are still not being created on this one.

File: public/js/draw/draw.js [hunk 8]
```diff
@@ -38,47 +38,127 @@ function getPosition(elem) {
     return { x: xPosition, y: yPosition };
 }
 
-
 /**
- * In node-mode creates a new node at the position of the click event on the SVG canvas.
- * In path-mode creates an elbow at the position of the click event on the SVG canvas,
-                if the startNode is defined.
- * @param {object} e The mousedown event.
+ * Creates a new node at position (posX, posY) on the SVG canvas.
```
Comment (david-yz-liu) [inline public/js/draw/draw.js]: Follow the documentation convention to properly document the function parameters.

File: public/js/draw/draw.js [hunk 9]
```diff
@@ -38,47 +38,127 @@ function getPosition(elem) {
...
+                code.setAttributeNS(null, 'y', textPosY);
+                code.setAttributeNS(null, 'class', 'mylabel'); // note: label is a class in bootstrap
+                var textNode = document.createTextNode(textStr);
+                code.appendChild(textNode);
+                g.appendChild(code);
+                document.getElementById(textId).addEventListener('mousedown', nodeClicked, false);
+            };
+        }
+        select(document.getElementById(nodeId_));
         nodeId += 1;
```
Comment (david-yz-liu) [inline public/js/draw/draw.js]: 1. I'm not sure what this line does if it's at the end of the function.
2. Are you missing a `}`? The indentation seems off.

Comment (R0ss) [inline public/js/draw/draw.js]: 1. `nodeId` is a global variable used to keep track of the next available ID, so it must be increment when used.

2. The indentation was off :+1:

File: public/js/draw/draw.js [hunk 10]
```diff
@@ -38,47 +38,131 @@ function getPosition(elem) {
...
+                code.setAttributeNS(null, 'align', textAlign);
+                code.setAttributeNS(null, 'x', textPosX);
+                code.setAttributeNS(null, 'y', textPosY);
+                code.setAttributeNS(null, 'class', 'mylabel'); // note: label is a class in bootstrap
+                var textNode = document.createTextNode(textStr);
+                code.appendChild(textNode);
+                g.appendChild(code);
+                document.getElementById(textId).addEventListener('mousedown', nodeClicked, false);
+            };
+        }
```
Comment (david-yz-liu) [inline public/js/draw/draw.js]: Okay, well now it looks like this entire block is not indented properly.

I recommend checking your indentation for the entire file again.

File: public/js/draw/draw.js [hunk 11]
```diff
@@ -38,47 +38,131 @@ function getPosition(elem) {
...
         node.inEdges = [];
         node.outEdges = [];
 
         g.appendChild(node);
         svgDoc.appendChild(g);
-        document.getElementById('n' + nodeId).addEventListener('mousedown', nodeClicked, false);
+        document.getElementById(nodeId_).addEventListener('mousedown', nodeClicked, false);
+
+        // Input Text into Shape
+        if (jsonObj){
```
Comment (Ian-Stewart-Binks) [inline public/js/draw/draw.js]: ` `

Comment (R0ss) [inline public/js/draw/draw.js]: Doh!

Comment (R0ss) [inline public/js/draw/draw.js]: There's gotta be a sublime default setting for this!

**Pull request merged**